### PR TITLE
A policy with 0 rules should return an error

### PR DESCRIPTION
**Pull request by ChiuWang**
**Which issue this PR fixes** 
[isuue#51565](https://github.com/kubernetes/kubernetes/issues/51565)

**Release note**: 
``` 
An audit policy file with 0 rule returns an error.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 1]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
 		return nil, err.ToAggregate()
 	}
 
-	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", len(policy.Rules), filePath)
+	lenth := len.(policy.Rules)
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Should be `len(policy.Rules)` instead of `len.(policy.Rules)`?

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: `sed s/lenth/length`

**Review by dixudx — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 2]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
 		return nil, err.ToAggregate()
 	}
 
-	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", len(policy.Rules), filePath)
+	s := len(policy.Rules)
```
Comment (soltysh) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Please use more meaningful names, rather than `s` here.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 3]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
...
+	if s < 1 {
+		return nil, fmt.Errorf("No audit policy rules loaded from file %s , check your configuration", filePath)
```
Comment (soltysh) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Errors should not start with uppercase, so `no audit...`

Comment (soltysh) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Maybe change the error to: `empty policy file is not allowed` or similar.

**Review by soltysh — APPROVED** [review]
A few nits, mostly lgtm. I'll defer to assigned reviewers for final call.

**Comment by CaoShuFeng** [comment]
This issue can't be fixed like this.
We should not return error for an empty file. Because empty file has 0 rules.
We should not return error for a policy file like this:
```
kind: Policy
apiVersion: audit.k8s.io/v1beta1
```
It also has 0 rules.


A file like this has 1 rule, but we can't load the rule from it.
```
# Log all requests at the Metadata level.
rules:
- level: Metadata
```
This is because no api-version is provided, so we can not decode it successfully.

**Comment by CaoShuFeng** [comment]
@ericchiang

**Comment by CaoShuFeng** [comment]
After some reconsideration, I have some others to say.
If we made an agreement that:
```
a policy file with 0 policy is illegal
```
Then this pr would be `GOOD`

What your opinion? @sttts @crassirostris @tallclair

**Comment by soltysh** [comment]
:+1: from me

**Comment by sttts** [comment]
> a policy file with 0 policy is illegal

:+1:

**Comment by CaoShuFeng** [comment]
@charrywanganthony is my workmate.
I will help him with this pr next Monday.
He has get off work today.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 4]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
 		return nil, err.ToAggregate()
 	}
 
-	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", len(policy.Rules), filePath)
+	length := len(policy.Rules)
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Nit: I think something like `policyCnt` is better for readability than `length`

**Review by crassirostris — COMMENTED** [review]
Could you please add a test for this case?

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 5]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
 		return nil, err.ToAggregate()
 	}
 
-	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", len(policy.Rules), filePath)
+	policyCnt := len(policy.Rules)
+	if policyCnt == 0 {
+		return nil, fmt.Errorf("failed to load policy rules from file %s", filePath)
+	}
+	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", policyCnt, filePath)
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: nit: glog messages shouldn't end in a `\n` (there is always an implicit newline).

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go [hunk 6]
```diff
@@ -49,6 +49,10 @@ func LoadPolicyFromFile(filePath string) (*auditinternal.Policy, error) {
 		return nil, err.ToAggregate()
 	}
 
-	glog.V(4).Infof("Loaded %d audit policy rules from file %s\n", len(policy.Rules), filePath)
+	policyCnt := len(policy.Rules)
+	if policyCnt == 0 {
+		return nil, fmt.Errorf("failed to load policy rules from file %s", filePath)
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: please make this error message more precise, e.g. `loaded illegal policy with 0 rules from file %s`

**Review by tallclair — CHANGES_REQUESTED** [review]
This needs a release note since it's technically a breaking change.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 7]
```diff
@@ -127,3 +138,32 @@ func TestParserV1beta1(t *testing.T) {
 		t.Errorf("Unexpected policy! Diff:\n%s", diff.ObjectDiff(policy, expectedPolicy))
 	}
 }
+
+func TestPolicyCntCheckV1alpha1(t *testing.T) {
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: I think we should parameterize the tests instead of duplicating the logic

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 8]
```diff
@@ -71,6 +72,16 @@ rules:
   - level: Metadata
 `
 
+const illegalPolicyDefV1alpha1 = `
+apiVersion: audit.k8s.io/v1beta1
```
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Nit: s/v1beta1/v1alpha1/

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 9]
```diff
@@ -71,6 +72,16 @@ rules:
...
+kind: Policy
+`
+
+const illegalPolicyDefV1beta1 = `
+apiVersion: audit.k8s.io/v1beta1
+kind: Policy
+`
+
```
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Adding such an illgeal rule would be better:

```
rules:
  - level: None
```

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 10]
```diff
@@ -127,3 +135,24 @@ func TestParserV1beta1(t *testing.T) {
 		t.Errorf("Unexpected policy! Diff:\n%s", diff.ObjectDiff(policy, expectedPolicy))
 	}
 }
+
+func TestPolicyCntCheck(t *testing.T) {
+
+	test := func(policyFileName string) {
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Try this approach instead of creating a function: https://golang.org/doc/code.html#Testing

Also, you can use the following patter to run test cases:

```
for _, tc := testCases {
  t.Run(tc.desc, func() {
    ...
  })
}
```

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 11]
```diff
@@ -127,3 +135,24 @@ func TestParserV1beta1(t *testing.T) {
...
+	        f, err := ioutil.TempFile("", "illegalPolicy.yaml")
+		require.NoError(t, err)
+		defer os.Remove(f.Name())
+
+		_, err = f.WriteString(policyFileName)
+		require.NoError(t, err)
+		require.NoError(t, f.Close())
+
+		_, err = LoadPolicyFromFile(f.Name())
+		assert.EqualError(t, err, fmt.Sprintf("loaded illegal policy with 0 rules from file %s", f.Name()))
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: I don't think copy-pasting the error to compare to is a good idea. If you change the message, you will have to change this place also. IMHO, a presence check will be enough

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go (hunk 10 above)
Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: I'll study it

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go (hunk 6 above)
Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go]: Done, plz check ^_^

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go (hunk 11 above)
Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Done, plz check ^_^

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 12]
```diff
@@ -32,7 +32,7 @@ import (
 )
 
 const policyDefV1alpha1 = `
-apiVersion: audit.k8s.io/v1beta1
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: why this change?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: disregard. This should be alpha.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 13]
```diff
@@ -71,6 +71,13 @@ rules:
   - level: Metadata
 `
 
+var testCases = []struct {
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Since this is a set of testCases for a specific function, it's better to move it inside this function

Sorry for not mentioning this earlier

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 14]
```diff
@@ -71,6 +71,13 @@ rules:
...
+	caseName, caseContent string
+}{
+	{`policyWithNoRule`, "apiVersion: audit.k8s.io/v1beta1\nkind: Policy"},
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: I think for readability it's better to avoid manually writing '\n'. Maybe do something like this instead?

```
{
  "policyWithNoRule",
  `
apiVersion: audit.k8s.io/v1beta1
kind: Policy
`,
}
```

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 15]
```diff
@@ -127,3 +134,19 @@ func TestParserV1beta1(t *testing.T) {
...
+		// Create an illegal policy file.
+		f, err := ioutil.TempFile("", "illegalPolicy.yaml")
+		require.NoError(t, err)
+		defer os.Remove(f.Name())
+
+		_, err = f.WriteString(tc.caseContent)
+		require.NoError(t, err)
+		require.NoError(t, f.Close())
+
+		_, err = LoadPolicyFromFile(f.Name())
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: This part is a copy-paste from another test. Extract to a separate function maybe?

Smth like `_, err := loadPolicyFromString(tc.policy)`

**Comment by sttts** [comment]
Removed label, waiting for @crassirostris's comments to be addressed.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go (hunk 15 above)
Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Thx for review. but, LoadPolicyFromFile is the function what I'm testing in reader.go.

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: I don't mind that, you will have the call to that function in the code. But if you want to have it called explicitly in the code -- that's OK also, wrap the rest at least:

```
f, err := writePolicy(tc.policy)
require.NoError(t, err)
defer os.Remove(f)

_, err = LoadPolicyFromFile(f)
```

Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: WriteString is a function imported from package "os" 
> func (f *File) WriteString(s string) (n int, err error)
maybe I can only change caseContent to policy?

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: > WriteString is a function imported from package "os"

I don't understand how it makes extracting a function impossible. Could you please elaborate?

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: What I mean is doing the following:

```
func TestPolicyCntCheck(t *testing.T) {
	for _, tc := range testCases {
		f, err := writePolicy(tc.policy)
		require.NoError(t, err)
		defer os.Remove(f)

		_, err = LoadPolicyFromFile(f)
		assert.Errorf(t, err, "loaded illegal policy with 0 rules from testCase %s", tc.caseName)
	}
}

func writePolicy(policy string) (string, error) {
		...
}
```

And replace the same piece of code in `TestParserV1beta1`

Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Oh very sorry about that. I misunderstood what you meant just now.

**Comment by CaoShuFeng** [comment]
LGTM.
But waiting for @crassirostris to label this pr.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 16]
```diff
@@ -110,20 +105,49 @@ func TestParserV1alpha1(t *testing.T) {
...
 	require.NoError(t, err)
 
 	assert.Len(t, policy.Rules, 3) // Sanity check.
 	if !reflect.DeepEqual(policy, expectedPolicy) {
 		t.Errorf("Unexpected policy! Diff:\n%s", diff.ObjectDiff(policy, expectedPolicy))
 	}
 }
+
+func TestPolicyCntCheck(t *testing.T) {
+	//a set of testCases
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: This comment adds no new information => not needed

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go [hunk 17]
```diff
@@ -110,20 +105,49 @@ func TestParserV1alpha1(t *testing.T) {
...
+		f, err := writePolicy(tc.policy, t)
+		require.NoError(t, err)
+		defer os.Remove(f)
+
+		_, err = LoadPolicyFromFile(f)
+		assert.Errorf(t, err, "loaded illegal policy with 0 rules from testCase %s", tc.caseName)
+	}
+}
+
+func writePolicy(policy string, t *testing.T) (string, error) {
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Small nit: usually `*testing.T ` comes as a first argument

**Review by crassirostris — COMMENTED** [review]
Several nits, overall LGTM!

Thanks a lot for your work!

**Comment by crassirostris** [comment]
Let's leave these nits for later

/lgtm

File: staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go (hunk 17 above)
Comment (charrywanganthony) [reply staging/src/k8s.io/apiserver/pkg/audit/policy/reader_test.go]: Thx for your careful instruction! I learned a lot.^_^
